### DISCUSSION TITLE
Failure initialization of SX1276 connected to Raspberry Pi 4

### DISCUSSION THREAD

**Discussion by simone-s** [jgromes/RadioLib#788] [Q&A]
Dear All,

we have the following configuration:

- Raspberry Pi 4 
- LoRa module SX1276

they are connected as follows:

```
LoRa SX 1276     |       RPi4
----------------------------- 
	3.3V		Pin 1
	GND		Pin 9
	MISO		Pin 21
	MOSI		Pin 19
	SCK		Pin 23
	NSS		Pin 22
	RESET		Pin 11
	DIO0		Pin7
```

with this configuration, we are assuming to use SPI channel 0 on the RPi4.

We have started with your example:
https://github.com/jgromes/RadioLib/tree/master/examples/NonArduino/Raspberry

We modified the main.cpp as follows to accommodate the hardware configuration above:

```
...

// create a new instance of the HAL class
// use SPI channel 1, because on Waveshare LoRaWAN Hat,
// the SX1261 CS is connected to CE1
PiHal* hal = new PiHal(0);

// now we can create the radio module
// pinout corresponds to the Waveshare LoRaWAN Hat
// NSS pin:   7
// DIO1 pin:  17
// NRST pin:  22
// BUSY pin:  not connected
SX1276 radio = new Module(hal, 24, 27, 11, RADIOLIB_NC);				

...
```

This code compiles directly on our target, i.e., RPi.
Unfortunately, this example program fails while running the initialization phase reporting the ERROR NUMBER -2.

We are using the last version of the RadioLib available on this repository.

May we ask for your support in using RadioLib with our hardware configuration?

We are also interested in using RadioLib to set ASK or FSK modulation on this SX1276 module. Is it possible?

Thank you in advance for your support.

**Comment by jgromes** [comment]
I'm a little bit confused about your pinout: in the first table, you seem to be referring to the physical pin number on the pin header. However, some of them seem wrong
* NSS: physical pin 22 is GPIO25, not 24 
* RST: physical pin 11 is GPIO17, not 11
* DIO0: physical pin 7 is GPIO4, not 27

So without knowing your hardware, I'm guessing the initialization should look like this:

```c++
SX1276 radio = new Module(hal, 25, 17, 4, RADIOLIB_NC);	
```

> we are assuming to use SPI channel 0 on the RPi4.

You have connected the NSS to a random GPIO pin, not one of the SPI0_CEx pins (GPIO7 or GPIO8), so I don't know what you mean by that. I don't know how the pigpio library behaves when using something else than CEx for the NSS pin.

> We are also interested in using RadioLib to set ASK or FSK modulation on this SX1276 module. Is it possible?

SX127x cannot do ASK, it can do LoRa, FSK or OOK, all of which are supported by RadioLib.

**Comment by simone-s** [comment]
Dear,

thank you for your answer.
We changed the NSS connection between the LoRa SX1276 and the RPi4.
Now we have

```

LoRa SX 1276     |       RPi4
----------------------------- 
	3.3V		Pin 1
	GND		Pin 9
	MISO		Pin 21
	MOSI		Pin 19
	SCK		Pin 23
	NSS		Pin 24
	RESET		Pin 11
	DIO0		Pin 7
```

We are now using the following initialization:

```
SX1276 radio = new Module(hal, 8, 17, 4, RADIOLIB_NC);	
```
Indeed, assuming that inside the Module function we should use the GPIOXX value and not the PINXX one, we connected the NSS to the GPIO8 (pin 24).
As reference we follow this pinout for our RPi4: https://pinout.xyz/#

We recompiled the NonArduino example code and something changed.
Now the example program fails while running the initialization phase reporting the ERROR NUMBER -16.

Do you have any suggestion?

Thanks.

**Comment by jgromes** [comment]
I would suggest to follow the issue template, at least by providing the debug + verbose output: https://github.com/jgromes/RadioLib/wiki/Debug-mode

-16 is SPI write failed, meaning the value in SPI register did not update even though it should have. Debug mode will show you exactly which register has failed.

**Comment by simone-s** [comment]
We recompiled in debug-mode and we got the follows

```
[SX1261] Initializing ... R	42	12	
M	SX127x
R	1	8	
R	1	F	
W	1	9	
R	1	8	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	8	
R	1	B	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	
R	1	F	

address:	0x1
bits:		2 0
value:		0x1
current:	0xF
mask:		0x7
new:		0x9
read:		0xF
failed, code -16
```

**Comment by simone-s** [comment]
Dear,

the first question about the Raspberry example. 
Since we do not have the Waveshare LoraWAN, do we have to drive the GPIO18 in PiHAL.h?
Because that pin should be the SPI1 CE0.

```
  void init() override {
      // first initialise pigpio library
      gpioInitialise();

      // now the SPI
      spiBegin();

      // Waveshare LoRaWAN Hat also needs pin 18 to be pulled high to enable the radio
      gpioSetMode(18, PI_OUTPUT);
      gpioWrite(18, PI_HIGH);
    }

    void term() override {
      // stop the SPI
      spiEnd();

      // pull the enable pin low
      gpioSetMode(18, PI_OUTPUT);
      gpioWrite(18, PI_LOW);

      // finally, stop the pigpio library
      gpioTerminate();
    }

```
Second, now we are initializing the LoRa module with this. 

``` 
SX1276 radio = new Module(hal, 8, 17, 4, RADIOLIB_NC);	

```
and we receive on the RPi4 terminal:

```
[SX1261] Initializing ... M	SX127x
Symbol length: 0.008000 ms
Symbol length: 4.096000 ms
success!
[SX1261] Transmitting packet ... failed, code -5
[SX1261] Transmitting packet ... failed, code -5

```

Is it correct? We use only one module, so we don't have any receivers.

Thanks.

**Comment by jgromes** [comment]
> Since we do not have the Waveshare LoraWAN, do we have to drive the GPIO18 in PiHAL.h?

as the comment says, this is something that's specific to Waveshare LoRaWAN hat, so no need for you to do that.

Next, error code -5 is transmit timeout, which likely means you do not have DIO0 connected.

It seems like this is not an issue in the library, so I'm converting it to a discussion instead.

**Comment by simone-s** [comment]
> as the comment says, this is something that's specific to Waveshare LoRaWAN hat, so no need for you to do that.

do we have to comment on those lines?

> Next, error code -5 is transmit timeout, which likely means you do not have DIO0 connected.

We connected DIO0 to GPIO7 = pin 7, is it correct?

And I do not completely understand how to connect the RESET. I'm sorry can you clarify this point as well?

Thanks.

**Comment by simone-s** [comment]
dear @jgromes,

We are fine with the fact that you changed this thread in a discussion.
The problem is that there are still some questions that we think are related to the usage of Radiolib that we didn't completely understand; see our previous message.

Thank you in advance for your support.

Kind regards.

**Comment by jgromes** [comment]
> do we have to comment on those lines?

You're not using it since your harwdare is different, so you can safely remove it.

> We connected DIO0 to GPIO7 = pin 7, is it correct?

It isn't, if you have initialized the module like you posted (`SX1276 radio = new Module(hal, 8, 17, 4, RADIOLIB_NC);`), then you should be using GPIO17 for DIO0. Also, GPIO7 is not physical pin 7, it's physical pin 26. You seem to be confusing physical pin numbers of the pin header with Raspberry GPIO pin numbers.

> And I do not completely understand how to connect the RESET. I'm sorry can you clarify this point as well?

Going off of the initialization above, you should be using GPIO4. To sum up:

`SX1276 radio = new Module(hal, 8, 17, 4, RADIOLIB_NC);` means:
* NSS connected to GPIO8
* DIO0 connected to GPIO17
* RST connected to GPIO4
* DIO1 not connected

I would suggest taking a look at the examples and documentation again, as well as to examine the difference between the pin header numbering and Raspberry Pi GPIO pin numbers.

**Comment by simone-s** [comment]
Dear @jgromes,

Thanks for your support.
Now it works, and we can transmit. The problem was in the wrong cabling between  RPi and LoRa modules.

Using the same NonArduino example for Raspberry on two "RPi+LoRa," unfortunately, we are not able to receive the "Hello World"" message. We got ERROR -24.

Thanks.

**Reply by jgromes** [reply]
As shown in the documentation here: https://jgromes.github.io/RadioLib/group__status__codes.html

-24 is `RADIOLIB_ERR_LORA_HEADER_DAMAGED`, this usually means that the transmitter has CRC disabled and receiver enabled. Can you post the code you are using for both transmitter and receiver?

**Reply by simone-s** [reply]
Hi,
We set `radio.setRCR(false) ` on both devices, but we still do not receive any message.

And we no longer have the error -24, but we don't understand if we receive any messages.

Thank you for any suggestions.

**Reply by jgromes** [reply]
Can you show how you initialize both ends of the link?
